pre_commit/clientlib.py

```diff
@@ -112,6 +112,34 @@ def validate_manifest_main(argv: Optional[Sequence[str]] = None) -> int:
 META = 'meta'
 
 
+class CheckMutableRev:
+    key = 'rev'
+
+    def check(self, dct: Dict[str, Any]) -> None:
+        if dct.get('repo') in {LOCAL, META}:
+            return cfgv.Conditional(
+                self.key,
+                cfgv.check_string,
+                condition_key='repo',
+                condition_value=cfgv.NotIn(LOCAL, META),
+                ensure_absent=True,
+            ).check(dct)
+
+        if any(
+            mutable in dct.get(self.key, '')
+            for mutable in {'HEAD', 'stable', 'master'}
+        ):
+            logger.warning(
+                f'The {self.key!r} field of repo {dct["repo"]!r} '
+                f"shouldn't be mutable. See https://pre-commit.com/#using-the-"
+                f'latest-version-for-a-repository for further details.',
+            )
+
+    apply_default = cfgv.Conditional.apply_default
+
+    remove_default = cfgv.Conditional.remove_default
```

Comment:
I'd still rather see this inherit from `Conditional` as I plan to delete `MigrateShaToRev` in pre-commit 3.x

Reply:
i had to change the inheritance to `ConditionalOptional`, due to the support of `sha` instead of `rev`:

```console
FAILED tests/commands/migrate_config_test.py::test_migrate_config_sha_to_rev - pre_commit.clientlib.InvalidConfigError:
```

also added a comment that the inheritance can (and should) be changed to `Conditional` if support for `sha` (e.g. deletion of `MigrateShaToRev`) is dropped.